src/schemathesis/models.py

```diff
@@ -189,6 +189,17 @@ def validate_response(
         if errors:
             raise AssertionError(*errors)
 
+    def __deepcopy__(self, memo: Optional[Dict[Any, Any]]) -> "Case":
+        return Case(
```

Comment:
Hm, is it intentional? As far as I understand, the first hook will modify the data, then the second hook will receive its modified version, which is probably not what we want - I assume that all these cases should be independent. Please, let me know if my assumption is correct

Reply:
> I assume that all these cases should be independent.

Yes, that is correct. I added [this test](https://github.com/kiwicom/schemathesis/pull/590/files#diff-3a837d2bf3cc88920c7ba962d6d94138R834) to ensure that mutating the Case object in place does not affect other cases. 

Before I defined `__deepcopy__` in the Case class, `deepcopy(case)` resulted in the exception `Expecting property name enclosed in double quotes' in "TypeError: cannot serialize '_io.FileIO' object`. I think this exception occurs trying to `deepcopy` the `form_data` field.

Reply:
@Stranger6667 this `deepcopy` problem is proving to be quite complicated. The sources of the `TypeError` are `Case.endpoint.app` and `Case.endpoint.schema.app`.

My current solution is to add a `__deepcopy__` function to the `Endpoint` class. This function does a `deepcopy` on most of the fields, shallow copies `Endpoint.app`, and uses `Endpoint.schema.clone()` (a shallow copy) to copy the `BaseSchema`. I will push this code for you to review.

The problem with this approach is that this is not a true deep copy yet this will be the generic way of deep copying the `Endpoint` class. If someone contributes code in the future that uses `deepcopy`, they may "successfully" and unknowingly get a partial deep copy of the `Endpoint`. The positive of this implementation is that it deep copies every field that we would reasonably expect the user to change.

The alternative approach is to implement a:
    
    class Case:
    ...
    def partial_deepcopy() -> "Case":
        endpoint_copy = Endpoint(
            headers=deepcopy(self.path_parameters),
            cookies=deepcopy(self.cookies),
            ...
            app=self.app
            schema=self.schema.clone()
        )
        return Case(
            endpoint=endpoint_copy
            path_parameters=deepcopy(self.path_parameters)
            ...
            app=self.app
        )

I think this `partial_deepcopy` function is probably better because it prevents any obscure logic errors using `deepcopy` on an Endpoint in the future.

Reply:
> The sources of the TypeError are Case.endpoint.app and Case.endpoint.schema.app.
Yep, generally, it is possible to re-construct such objects, but I don't think it worth the efforts. Basically it was one of the reasons I removed these things from "serialized" events that are yielded by the runner.

I think that the `partial_deepcopy` implementation is the way to go :) 

:+1: